apollo-router/tests/type_conditions_test.rs

```diff
@@ -0,0 +1,182 @@
+                },
+                {
+                    "__typename":"ArticleResult",
+                    "id":"5092bbea-8bc3-4c4f-a9eb-003604ed9add",
+                    "sections":[
+                        {
+                            "__typename":"GallerySection",
+                            "id":"798e75ae-9378-41de-a014-af9f9a5e99eb"
+                        },
+                        {
+                            "__typename":"GallerySection",
+                            "id":"f756501a-7377-4081-861b-0097cbfb7f41"
+                        }
+                    ]
+                }
+            ]
+        }
+    }}).build();
+
+    let artwork_service = MockSubgraph::builder().with_json(json!{{
+        "query":"query Search__artworkSubgraph__1($representations:[_Any!]!){_entities(representations:$representations){...on EntityCollectionSection{artwork title}...on GallerySection{artwork}}}","operationName":"Search__artworkSubgraph__1",
+        "variables":{
+            "representations":[
+                {
+                    "__typename":"EntityCollectionSection",
+                    "id":"a7487f33-bd37-48a6-a843-c0cda86f5049"
+                },
+                {
+                    "__typename":"EntityCollectionSection",
+                    "id":"cdb43f1d-df2d-4293-a328-8d38a0cdd742"
+                },
+                {
+                    "__typename":"GallerySection",
+                    "id":"798e75ae-9378-41de-a014-af9f9a5e99eb"
+                },
+                {
+                    "__typename":"GallerySection",
+                    "id":"f756501a-7377-4081-861b-0097cbfb7f41"
+                }
+            ]
+        }
+    }},
+json!{{
+        "data":{
+            "_entities":[
+                {
+                    "artwork":"Hello World",
+                    "title":"Hello World"
+                },
+                {
+                    "artwork":"Hello World",
+                    "title":"Hello World"
+                }
+            ]
+        }
+    }}).build();
+
+    let mut mocks = MockedSubgraphs::default();
+    mocks.insert("searchSubgraph", search_service);
+    mocks.insert("artworkSubgraph", artwork_service);
+
+    let schema = include_str!("fixtures/type_conditions.graphql");
+    TestHarness::builder()
+        .try_log_level("info")
+        .configuration_json(configuration)
+        .unwrap()
+        .schema(schema)
+        .extra_plugin(mocks)
+}
+
+static QUERY: &str = "
+query Search {
+    search {
+      ... on MovieResult {
+        sections {
+          ... on EntityCollectionSection {
+            artwork
+            id
+            title
+          }
+          ... on GallerySection {
+            artwork
+            id
+          }
+        }
+        id
+      }
+      ... on ArticleResult {
+        id
+        sections {
+          ... on GallerySection {
+            artwork
+          }
+          ... on EntityCollectionSection {
+            artwork
```

Comment:
to make it more obvious you could pass a value that is specific to the type condition and that value back from the subgraph.
```graphql
artwork(params: "articleResult")
```
and expect the value to be returned as `"blah-articleResult"`

Reply:
Great idea, thanks!